Add a guide on xs-leaks (#38977)
mdn/content: files/en-us/web/security/attacks/xs-leaks/index.md

---
title: Cross-site leaks (XS-leaks)
slug: Web/Security/Attacks/XS-leaks
page-type: guide
---

> Cross-site leaks (also called XS-leaks) are a class of attack which exploit the ways in which a target site can leak information to an attacker's site, either through web platform APIs or through {{glossary("side channels")}} such as timing attacks. The information leaked could include, for example:

Reviewer: I think a web platform API is a side channel in these cases (i.e. is it worth differentiating)?

Reviewer 2: I think the wording is a little unclear: "a class of attack which exploit the ways in which a target site can leak information to an attacker's site". Maybe start with a typical user story, an example of what information is (presumably unintentionally) leaking and then how an attacker exploits that leak?

Author: > I think a web platform API is a side channel in these cases (i.e. is it worth differentiating)?

yes, tbh I started being uncertain about what exactly counts as a side channel, so I reworked the definition a bit in https://github.com/mdn/content/pull/38977/commits/36d919efdac4826b7fbc47630e26471a912e894b .

> Maybe start with a typical user story, an example of what information is (presumably unintentionally) leaking and then how an attacker exploits that leak?

I'm happy to do more work on the definition if it is unclear but I really do want to start with a concise definition rather than a story.

- Whether the user has visited the target site.
- Whether the user is logged into the target site.
- What the user's ID on the site is.

> - What the user has recently searched for on the site.

Reviewer: It is worth a following paragraph as to why a user or developer should care if any of this leaks, and perhaps some specific dangers from this knowledge [I had the same problem with the discussion of timing attacks on the wiki page - to motivate change people have to understand not just what the risk is, but the impact of the risk]. 

My first reaction was that I don't see the risk. On further thought, I guess if I know that the user has a login with bank say, that makes them a better target for trying some of the other attacks, or even an old-style scam where I send a letter - but in this case I know that they will be in that bank, and I might know their id, which adds to the trustworthyness.

Further, if I know their id, perhaps that allows a dark web search for passwords etc. Dunno, but worth making it real world.

Author: > It is worth a following paragraph as to why a user or developer should care if any of this leaks, and perhaps some specific dangers from this knowledge [I had the same problem with the discussion of timing attacks on the wiki page - to motivate change people have to understand not just what the risk is, but the impact of the risk].
> 
> My first reaction was that I don't see the risk. On further thought, I guess if I know that the user has a login with bank say, that makes them a better target for trying some of the other attacks, or even an old-style scam where I send a letter - but in this case I know that they will be in that bank, and I might know their id, which adds to the trustworthyness.
> 
> Further, if I know their id, perhaps that allows a dark web search for passwords etc. Dunno, but worth making it real world.

It's a good point. AIUI these attacks are usually less severe than something like an XSS or even a CSRF attack. There are a couple of things worth mentioning:
- the mere fact that someone knows you have an active account somewhere could put you in peril, and certainly might be something you don't want people to know about you. It could open you up to retaliation from an oppressive government (for example, if you are seeking an abortion) or open you up to extortion.
- as you've noted, it would make a phishing attack (for example) much more convincing if someone knows your ID

Worth mentioning these I agree.

Author: -> https://github.com/mdn/content/pull/38977/commits/575018586eb90ef291a59889b6b3cbc46c0156da

> Unlike other attacks such as [XSS](/en-US/docs/Web/Security/Attacks/XSS) or [Clickjacking](/en-US/docs/Web/Security/Attacks/Clickjacking), cross-site leaks are not a single technique. Instead, they are a term for a whole class of attack which exploit weaknesses in the ways that browsers isolate websites from each other.

Reviewer: Do we need to say something about how these are typically deployed? I mean how do you get someone to visit evil.com - does it matter? [what I'm thinking here is that for some of the other attacks we said things like "send target an email that does X, Y]]

Author: -> https://github.com/mdn/content/pull/38977/commits/67f993d20f8ca01aaf6eb43f1f0019942b752a3e

In this guide we will not attempt to describe every cross-site leak attack and defense. Instead, we'll start by describing a few example attacks, then outline the common underlying weaknesses that enable them, then describe some general defenses that can work against many known attacks.

## Sample cross-site leaks

In this section we'll describe three different cross-site leaks, to give an idea of how they work.

- [Leaking page existence using error events](#leaking_page_existence_using_error_events): in this attack, an attacker can determine whether particular pages in the target site exist, by attempting to load them as resources and listening for the {{domxref("HTMLElement/error_event", "error")}} and {{domxref("HTMLElement/load_event", "load")}} events. If certain pages are only available to logged-in users, the attacker can determine whether the user is signed into the target site.

> - [Frame counting using window references](frame_counting_using_window_references): in this attack, the attacker gets a reference to a {{domxref("Window", "window")}} object hosting a page in the target site, for example as the return value of a call to {{domxref("Window.open()", "window.open()")}}. The attacker can then determine the number of {{htmlelement("iframe")}} elements in the target page, which again might reveal whether the user is signed into the target.

Reviewer: Just FMI - how can the number of iframes in a page reveal whether the user is logged in? I guess if the page displays different content for logged in users.

It might be worth saying something along the lines at the top level that the attacker does need to have some knowledge of the target site or you wouldn't know this many frames means logged in, or what page to test for your load event. 

Note, I'm not sure, this is one of those "what went though my head as I read this reviews". What went through my head here is that this was a very specific search about a user interaction with a particular site, and therefore the risk (and hence protection effort) kind of depends on what your site is hiding.

Author: This is all true, and yes, AIUI the idea is that the _if_ target include a a different number of iframes for logged-in users (which is very plausible), then the number of iframes gives that away, and so yes, this is a bespoke attack for a particular target (but many of these attacks are like that).

> - [Leaking redirects with a CSP](leaking_redirects_with_a_csp): in this attack, the attacker creates a [Content Security Policy](/en-US/docs/Web/HTTP/Guides/CSP) which effectively blocks redirects, then tries to load a page in the target site. If the page is blocked, the attacker knows that the target redirected the request, and again, this may reveal that the user is signed in.

Reviewer: ```suggestion
- [Leaking redirects with a CSP](#leaking_redirects_with_a_csp): in this attack, the attacker creates a [Content Security Policy](/en-US/docs/Web/HTTP/Guides/CSP) which effectively blocks redirects, then tries to load a page in the target site. If the page is blocked, the attacker knows that the target redirected the request, and again, this may reveal that the user is signed in.
```

> - [Leaking redirects with a CSP](leaking_redirects_with_a_csp): in this attack, the attacker creates a [Content Security Policy](/en-US/docs/Web/HTTP/Guides/CSP) which effectively blocks redirects, then tries to load a page in the target site. If the page is blocked, the attacker knows that the target redirected the request, and again, this may reveal that the user is signed in.

Reviewer: My understanding is that if a page tries to load a resource that is served with a CSP policy that the calling page violates, the browser simply doesn't load it, and fires a `securityviolation` event. What is this redirect we're talking about?

Author: The target doesn't have the CSP, the attacker does. The attacker creates a CSP - for its page - which only allows the target page to be loaded, then tries to load the target page.

> - [Leaking redirects with a CSP](leaking_redirects_with_a_csp): in this attack, the attacker creates a [Content Security Policy](/en-US/docs/Web/HTTP/Guides/CSP) which effectively blocks redirects, then tries to load a page in the target site. If the page is blocked, the attacker knows that the target redirected the request, and again, this may reveal that the user is signed in.

Reviewer: >  If the page is blocked, the attacker knows that the target redirected the request, and again, this may reveal that the user is signed in.

I don't understand how this inference can be drawn. Might be obvious in the section. Is it that some pages are only served with CSP for logged in users or something similar - so you won't get the violation report for a non-logged in user?

Reviewer: Actually, maybe you're saying that the logged in user will be taken to the trusted page, and IF this page is served with CSP there will be a violation. If they are not logged in the user will be taken to another page - a redirect has occurred. So if that redirect page is not protected by some kind of CSP, then you will be seeing different behaviour, which tells you something about state.

Author: The CSP is controlling the attacker page, not the target page.

See the example: https://pr38977.review.mdn.allizom.net/en-US/docs/Web/Security/Attacks/XS-leaks#leaking_redirects_with_a_csp. 

This is the attacker page, and it contains a CSP in the `<meta>` tag. The CSP says that this page (i.e. the attacker page) can only load the target page to be loaded into an `<iframe>`. The attacker page then tries to load the target page into an `<iframe>`.

- If this does not generate a `securitypolicyviolation` event, then the target did not redirect.
- If it does generate a `securitypolicyviolation` event, then the target did redirect.

The doc does say this:

> In this attack, the attacker creates a page with a [Content Security Policy (CSP)](https://pr38977.review.mdn.allizom.net/en-US/docs/Web/HTTP/Guides/CSP) that will generate a [securitypolicyviolation](https://pr38977.review.mdn.allizom.net/en-US/docs/Web/API/Element/securitypolicyviolation_event) event if the page tries to embed a document that is not from the target URL.

Reviewer: Yes, I see it now, but I needed this explanation to understand the code. I think sometimes I'm a bit dim, but I can't be along. Maybe we can improve it. https://github.com/mdn/content/pull/38977/files#r2032522590

Author: I'm happy to make the change you suggested, but I don't expect people to be able to understand the attack from the little intro paragraph at https://pr38977.review.mdn.allizom.net/en-US/docs/Web/Security/Attacks/XS-leaks#sample_cross-site_leaks - if I did, I wouldn't need the full description and code sample further down.

This `<ul>` is really just giving a preview of what we will talk about in the section. It might even be better to remove the explanation and just make it a list of fragment links:

> - [Leaking page existence using error events](https://pr38977.review.mdn.allizom.net/en-US/docs/Web/Security/Attacks/XS-leaks#leaking_page_existence_using_error_events)
> - [Frame counting using window references](https://pr38977.review.mdn.allizom.net/en-US/docs/Web/Security/Attacks/frame_counting_using_window_references)
> - [Leaking redirects with a CSP](https://pr38977.review.mdn.allizom.net/en-US/docs/Web/Security/Attacks/XS-leaks#leaking_redirects_with_a_csp)

script.src = "https://example.org/admin";
document.head.appendChild(script);
```

This results in an HTTP request to the `https://example.org/` website. If the request includes cookies that the site uses to identify users, and the page requested is only available to logged-in users, then the success or failure of the request reveals whether or not the user is logged in.

If the request fails, then the element fires an {{domxref("HTMLElement.error_event", "error")}} event. If the request succeeds, the element fires a {{domxref("HTMLElement.load_event", "load")}} event. By listening for these events, the attacker can discover whether the user is logged in.

```js
const url = "https://example.org/admin";
const script = document.createElement("script");

script.addEventListener("load", (e) => {
  console.log(`${url} exists`);
});

script.addEventListener("error", (e) => {
  console.log(`${url} does not exist`);
});

script.src = url;
document.head.appendChild(script);
```

An attacker may even be able to discover a user's ID, by iteratively trying to load pages to see if pages like `https://example.org/users/my_username` exist.

### Frame counting using window references

> In a frame counting attack, the attacker finds out the number of frames currently loaded in the target page. In turn, that leaks information about the state of the target site, which may enable to attacker to learn, for example, whether the user is currently logged into the site.

Reviewer: OK, so yes the attacker can draw inferences about a site when they know characteristics of the side that reflect certain states, such as logged in. IN this case the characteristic is the number of loaded frames.

Author: Well for example, someone might have special status that's more fine-grained than just "logged in", like platinum status on a banking site (whatever that means).

Reviewer: Or admin, as you showed in the other example.

If an attacker site gets a reference to a {{domxref("Window")}} object containing the target site, the attacker can count the number of frames in the target site by reading the {{domxref("Window.length", "window.length")}} property.

The attacker can get a `Window` object by calling {{domxref("Window.open()", "window.open()")}}:

```js
const target = window.open("https://example.org");
const frames = target.length;
```

Alternatively, the attacker can embed the target site in an {{htmlelement("iframe")}}, and retrieve the frame's {{domxref("HTMLIFrameElement.contentWindow", "contentWindow")}} property:

```html
<iframe src="https://example.org"></iframe>
```

```js
const target = document.querySelector("iframe").contentWindow;
const frames = target.length;
```

### Leaking redirects with a CSP

In this attack, the attacker creates a page with a [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/Guides/CSP) that will generate a {{domxref("Element.securitypolicyviolation_event", "securitypolicyviolation")}} event if the page tries to embed a document that is not from the target URL.

> The attacker's page then creates an {{htmlelement("iframe")}} that embeds the target URL, and listens for the violation event. If the violation event fires, the attacker knows that a redirect took place, and this in turn may reveal something about the user's relationship with the site (for example, that they are signed in, or that they have some other status on the site).

Reviewer: "attacker knows that a redirect took place" - see comment above on redirect https://github.com/mdn/content/pull/38977/files#r2030329489 - I'm not sure what is happening here and why is crystal clear - at least it isn't for me.

Author: So yes, the idea is that the attack leaks whether there was a redirect, and that lets the attacker know whether the user is signed in.

For instance:

- if I am signed into https://frontendmasters.com/ and I visit https://frontendmasters.com/dashboard/, there's no redirect - it just shows my dashboard.
- if I am not signed in, then it redirects me to https://frontendmasters.com/login/

But it's also not just signed in/not signed in of course. There might be some special status someone has on a site that this might leak.

Author: I've tried to break this down some more in https://github.com/mdn/content/pull/38977/commits/e04b3dedd67c478fb423169c545163789c52ad02.

Reviewer: Yes, this is really good. I like the use of the admin here and explanation - previously admin login was used in an example and it looked pointless. Now it it is clear that this tells you not just that they are logged in but that they are an admin - compromising this login is more worthwhile than others.

```html
<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="frame-src https://example.org" />
  </head>
  <body>
    <script>
      document.addEventListener("securitypolicyviolation", () => {
        console.log("Page was redirected");
      });
      const frame = document.createElement("iframe");
      document.body.appendChild(frame);
      frame.src = "https://example.org";
    </script>
  </body>
</html>
```

Note that this attack works even if the target site disallows embedding using a mechanism such as [`frame-ancestors`](/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/frame-ancestors).

## Defenses against cross-site leaks

Cross-site leaks exploit mechanisms in the web platform which enables websites to interact with each other. Correspondingly, the defenses against cross-site leaks all involve _isolating_ the target website from potential attackers, by disabling or controlling these cross-site interactions.

> Since cross-site leaks can work in many different ways, there isn't a single defense that will work against all of them. However, there are several practices that will work against many of them, and we will summarise them here.

Reviewer: Don't we use US spellings?
```suggestion
Since cross-site leaks can work in many different ways, there isn't a single defense that will work against all of them. However, there are several practices that will work against many of them, and we will summarize them here.
```

```js
// Attempt to load a page in the target as a resource
const script = document.createElement("script");
script.src = "https://example.org/admin";
document.head.appendChild(script);
```

A server can use Fetch metadata to deny these requests, as in the following [Express](/en-US/docs/Learn_web_development/Extensions/Server-side/Express_Nodejs) code:

```js
function isAllowed(req) {
  // Allow same-origin, same-site, and user-initiated requests
  const secFetchSite = req.headers["sec-fetch-site"];
  if (
    secFetchSite === "same-origin" ||
    secFetchSite === "same-site" ||
    secFetchSite === "none"
  ) {
    return true;
  }

  // Allow cross-site navigations, such as clicking links
  const secFetchMode = req.headers["sec-fetch-mode"];
  if (secFetchMode === "navigate" && req.method === "GET") {
    return true;
  }

  // Deny everything else
  return false;
}

app.get("/admin", (req, res) => {
  if (isAllowed(req)) {
    // Respond with the admin page if the user is admin
    getAdminPage(req, res);
  } else {
    res.status(404).send("Not found.");
  }
});
```

> 

Reviewer: So what specifically happens here that is different

```suggestion
With the above server code the page is never loaded: the problematic code will always get an error whether the user is logged in or not.
```

Author: -> https://github.com/mdn/content/pull/38977/commits/c854152cd32e9498d05a4db5dcd65887644dad70

A policy like this is called a _Resource Isolation Policy_. To learn much more about implementing isolation policies with Fetch metadata, see [Protect your resources from web attacks with Fetch Metadata](https://web.dev/articles/fetch-metadata) and [Isolation Policies](https://xsleaks.dev/docs/defenses/isolation-policies/).

### SameSite cookies

The [`SameSite`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#samesitesamesite-value) cookie attribute determines whether or not the cookie will be sent in requests originating from a different site.

The `Lax` value of `SameSite` means that cross-site requests will only include the cookie if the request is a top-level navigation (meaning, essentially, that the value in the browser's address bar changes to the target site) and uses a {{glossary("Safe/HTTP", "safe")}} method (most notably, this excludes {{httpmethod("POST")}} requests).

> This can protect against some cross-site leaks. For example, the [Leaking page existence using error events](#leaking_page_existence_using_error_events) attack depends on the attacker making cross-site resource requests that include the user's cookies. Setting `SameSite` to `Lax` on the user's session cookie would prevent this attack, because the cookie would not be included in the attacker's request, so the admin page would never be returned.

Author: I think 'even" should be "ever" here, because the point is that the page is not returned whether or not the user is logged in.

As a rule, `SameSite` should be treated as a defense in depth measure, and should be deployed as well as a more explicit isolation policy such as one based on Fetch metadata.

### Framing protection

Many cross-site leaks rely on the attacking site being able to embed the target as an {{htmlelement("iframe")}}. For example, this is one method an attacker can use to get a reference to the target's {{domxref("Window", "window")}}, to enable a [frame-counting](#frame_counting_using_window_references) attack.

This means it is a good practice to prevent a site from being embeddable unless you need to allow embedding, and if you do need to allow embedding, restrict it as far as you can.

There are two relevant tools here:

- The [`frame-ancestors` directive](/en-US/docs/Web/HTTP/CSP#clickjacking_protection) in a [content security policy](/en-US/docs/Web/HTTP/CSP).
- The {{httpheader("X-Frame-Options")}} response header.

> The `frame-ancestors` directive is a replacement for `X-Frame-Options`. By setting `X-Frame-Options` as well as `frame-ancestors`, you can prevent embedding in browsers that don't support `frame-ancestors`. As [browser support for `frame-ancestors` is very good](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors#browser_compatibility), this is not a major concern.
> 
> If `frame-ancestors` and `X-Frame-Options` are both set, then browsers that support `frame-ancestors` will ignore `X-Frame-Options`.

Reviewer: ```suggestion
The `frame-ancestors` directive is a replacement for `X-Frame-Options`. By setting `X-Frame-Options` as well as `frame-ancestors`, you can prevent embedding in browsers that don't support `frame-ancestors` (if both are both set, then browsers that support `frame-ancestors` will ignore `X-Frame-Options`). As [browser support for `frame-ancestors` is very good](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors#browser_compatibility), this is not a major concern.
```

### Cross-Origin Opener Policy (COOP)

As we saw in the [frame-counting](#frame_counting_using_window_references) attack, another way to get a reference to the target's {{domxref("Window", "window")}} is as the return value of a call to {{domxref("Window.open()", "window.open()")}}:

```js
const target = window.open("https://example.com");
```

The {{httpheader("Cross-Origin-Opener-Policy")}} response header determines whether a document will be opened in the same {{glossary("Browsing context", "browsing context group")}} as the document that opened it.

If your server sends this header and sets it to any value except the default of `"unsafe-none"`, then if a document from a different origin tries to open your page using `window.open()`, your page will be loaded into a different browsing context group. Among other things, this means that the opener will not get a reference to the `window` object for your page, and will therefore not be able to use it in a frame counting attack.

### Defense summary checklist

> As we've seen, cross-site leaks include a range of attacks targeting different parts of the web platform: a single defense doesn't work against any of them, and some leaks don't have any defenses yet.

Reviewer: > and some leaks don't have any defenses yet.

Pedantic, but we have not seen this in this doc.

Author: Actually the CSP one doesn't have any defenses AFAIK. Should probably say that.

Author: -> https://github.com/mdn/content/pull/38977/commits/beae2f35d07741d4946a9c1c380ff4d2f551a4f8

## Sample cross-site leaks

In this section we'll describe three different cross-site leaks, to give an idea of how they work.

- [Leaking page existence using error events](#leaking_page_existence_using_error_events): in this attack, an attacker can determine whether particular pages in the target site exist, by attempting to load them as resources and listening for the {{domxref("HTMLElement/error_event", "error")}} and {{domxref("HTMLElement/load_event", "load")}} events. If certain pages are only available to logged-in users, the attacker can determine whether the user is signed into the target site.
- [Frame counting using window references](frame_counting_using_window_references): in this attack, the attacker gets a reference to a {{domxref("Window", "window")}} object hosting a page in the target site, for example as the return value of a call to {{domxref("Window.open()", "window.open()")}}. The attacker can then determine the number of {{htmlelement("iframe")}} elements in the target page, which again might reveal whether the user is signed into the target.

> - [Leaking redirects with a CSP](#leaking_redirects_with_a_csp): in this attack, the attacker's page has a [Content Security Policy](/en-US/docs/Web/HTTP/Guides/CSP) which effectively blocks redirects, then tries to load a page in the target site. If the page is blocked, the attacker knows that the target redirected the request, and again, this may reveal that the user is signed in.

Author: I'm happy to make the change you suggested, but I don't expect people to be able to understand the attack from the little intro paragraph at https://pr38977.review.mdn.allizom.net/en-US/docs/Web/Security/Attacks/XS-leaks#sample_cross-site_leaks - if I did, I wouldn't need the full description and code sample further down.

This `<ul>` is really just giving a preview of what we will talk about in the section. It might even be better to remove the explanation and just make it a list of fragment links:

> - [Leaking page existence using error events](https://pr38977.review.mdn.allizom.net/en-US/docs/Web/Security/Attacks/XS-leaks#leaking_page_existence_using_error_events)
> - [Frame counting using window references](https://pr38977.review.mdn.allizom.net/en-US/docs/Web/Security/Attacks/frame_counting_using_window_references)
> - [Leaking redirects with a CSP](https://pr38977.review.mdn.allizom.net/en-US/docs/Web/Security/Attacks/XS-leaks#leaking_redirects_with_a_csp)

> - [Leaking redirects with a CSP](#leaking_redirects_with_a_csp): in this attack, the attacker's page has a [Content Security Policy](/en-US/docs/Web/HTTP/Guides/CSP) which effectively blocks redirects, then tries to load a page in the target site. If the page is blocked, the attacker knows that the target redirected the request, and again, this may reveal that the user is signed in.

Author: Hm, I just noticed that your suggestion added a new paragraph, I don't really like this here, I think it slows things down to go into too much detail about the attacks at this point. Maybe a bare list as suggested above is better.

> - [Leaking redirects with a CSP](#leaking_redirects_with_a_csp): in this attack, the attacker's page has a [Content Security Policy](/en-US/docs/Web/HTTP/Guides/CSP) which effectively blocks redirects, then tries to load a page in the target site. If the page is blocked, the attacker knows that the target redirected the request, and again, this may reveal that the user is signed in.

Reviewer: FWIW I like the way it is now. The previous explanation confused me, and this does not.

---
title: Cross-site leaks (XS-Leaks)
slug: Web/Security/Attacks/XS-Leaks
page-type: guide
sidebar: security
---

Cross-site leaks (also called XS-Leaks) are a class of attack in which an attacker's site can derive information about the target site, or about the user's relationship with the target site, by using web platform APIs that enable sites to interact with one another. The information leaked could include, for example:

- Whether the user has visited the target site.
- Whether the user is logged into the target site.
- What the user's ID on the site is.
- What the user has recently searched for on the site.

This might seem to be a much less damaging problem than, for example, a [cross-site scripting](/en-US/docs/Web/Security/Attacks/XSS) attack, but it can still have serious consequences for users. For example:

> - A user might have accounts on websites that they don't want to make public. Leaking this information to an attacker could expose them to extortion or retaliation from an oppressive government (for example, against a user seeking information about abortions).

Reviewer 3: Could we rephrase this?

```suggestion
- A user might have accounts on websites that they don't want to make public. Leaking this information to an attacker could expose them to extortion or retaliation from an oppressive government (for example, against a user seeking information about specific medical procedures).
```

```html
<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="frame-src https://example.org/admin" />
  </head>
  <body>
    <script>
      document.addEventListener("securitypolicyviolation", () => {
        console.log("Page was redirected");
      });
      const frame = document.createElement("iframe");
      document.body.appendChild(frame);
      frame.src = "https://example.org/admin";
    </script>
  </body>
</html>
```

- If the user is logged in as an admin, then the `<iframe>` will load, and the browser will not fire `securitypolicyviolation`.
- If the user is not logged in as admin, the server redirects to `https://example.org/login`. Because this URL is not allowed by the attacker's CSP, the browser will block the `<iframe>` and fire the `securitypolicyviolation` event, and the attacker's event handler will run.

Note that this attack works even if the target site disallows embedding using a mechanism such as [`frame-ancestors`](/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/frame-ancestors).

## Defenses against cross-site leaks

> Cross-site leaks exploit mechanisms in the web platform which enables websites to interact with each other. Correspondingly, the defenses against cross-site leaks all involve _isolating_ the target website from potential attackers, by disabling or controlling these cross-site interactions.

Reviewer 4: > the defenses against cross-site leaks all involve

Not all protections involve isolating the target. For example, adding XSRF tokens to subresources do not isolate the target but prevent an attacker from probing these resources.

Author: -> https://github.com/mdn/content/pull/38977/commits/ccf9aa32ecf5abddbf518ec386369d40d34fe221

- {{httpheader("Sec-Fetch-Mode")}}: The request's {{domxref("Request.mode", "mode")}}.
- {{httpheader("Sec-Fetch-User")}}: Whether the request is a user-initiated navigation.
- {{httpheader("Sec-Fetch-Dest")}}: The request's {{domxref("Request.destination", "destination")}}.

Fetch metadata headers are not a defense mechanism on their own, but enable a server to implement a policy which will deny requests that are used in cross-site leaks as well as other attacks such as [Cross-Site Request Forgery (CSRF)](/en-US/docs/Web/Security/Attacks/CSRF) attacks.

For example, the [Leaking page existence using error events](#leaking_page_existence_using_error_events) attack depends on the attacker being able to make cross-site requests to load, as resources, pages that belong to the target:

```js
// Attempt to load a page in the target as a resource
const script = document.createElement("script");
script.src = "https://example.org/admin";
document.head.appendChild(script);
```

A server can use Fetch metadata to deny these requests, as in the following [Express](/en-US/docs/Learn_web_development/Extensions/Server-side/Express_Nodejs) code:

```js
function isAllowed(req) {

>   // Allow same-origin, same-site, and user-initiated requests

Reviewer 4: "none" are either user initiated or browser initiated, _I think_.

Author: Could you elaborate what the difference is? The spec only talks about user-initiated (https://w3c.github.io/webappsec-fetch-metadata/#sec-fetch-site-header)  and links to come more detail at https://w3c.github.io/webappsec-fetch-metadata/#directly-user-initiated.

Reviewer 4: Didn't know that the spec tries to be that explicit about "none" and I'm surprised there is no mention about browser initiated requests but maybe I'm hallucitating. I thought that some of the background requests the browser sends come with `Sec-Fetch-Site: none` but maybe the headers are just omitted in that scenario. Let's leave it how it is for now :)

This means it is a good practice to prevent a site from being embeddable unless you need to allow embedding, and if you do need to allow embedding, restrict it as far as you can.

There are two relevant tools here:

- The [`frame-ancestors` directive](/en-US/docs/Web/HTTP/CSP#clickjacking_protection) in a [content security policy](/en-US/docs/Web/HTTP/CSP).
- The {{httpheader("X-Frame-Options")}} response header.

The `frame-ancestors` directive is a replacement for `X-Frame-Options`. Although [browser support for `frame-ancestors` is very good](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors#browser_compatibility), some very old browsers, notably Internet Explorer, don't support `frame-ancestors`.

If `frame-ancestors` and `X-Frame-Options` are both set, then browsers that support `frame-ancestors` will ignore `X-Frame-Options`. This means that there's no reason not to set `X-Frame-Options` as well as `frame-ancestors`, and thus prevent embedding even in browsers that don't support `frame-ancestors`.

> ### Cross-Origin Opener Policy (COOP)

Reviewer 4: I would maybe suggest mentioning that COOP in many cases limits the attacks to 1 bit of information per user activation because the reference is "burned" after COOP headers are encountered. Example 1 bit attacks that are possible with COOP are timing based attacks - you can measure the time it takes to lose reference to the window.

Author: I wonder if this is a bit too advanced/nuanced for this guide.

Reviewer 4: From the experience COOP cause a little confusion because it's presented as a complete mitigation against navigation based xsleaks but it's a great to mitigate the risks rather than fully mitigate them. It is a little bit advanced/nuanced so it's your call in the end whether it's worth mentioning the nuance :)

---
title: Cross-site leaks (XS-leaks)
slug: Web/Security/Attacks/XS-leaks
page-type: guide

> ---

Reviewer 3: ```suggestion
sidebar: security
---
```

- A user might have accounts on websites that they don't want to make public. Leaking this information to an attacker could expose them to extortion or retaliation from an oppressive government (for example, against a user seeking information about abortions).
- Knowing a user has an account on a site, especially if their user ID can be determined, can make a subsequent phishing attack much more convincing.

Unlike other attacks such as [XSS](/en-US/docs/Web/Security/Attacks/XSS) or [Clickjacking](/en-US/docs/Web/Security/Attacks/Clickjacking), cross-site leaks are not a single technique. Instead, they are a term for a whole class of attack which exploit weaknesses in the ways that browsers isolate websites from each other.

In this guide we will not attempt to describe every cross-site leak attack and defense. Instead, we'll start by describing a few example attacks, then outline the common underlying weaknesses that enable them, then describe some general defenses that can work against many known attacks.

## Sample cross-site leaks

In this section we'll describe three different cross-site leaks, to give an idea of how they work.

> - [Leaking page existence using error events](#leaking_page_existence_using_error_events): in this attack, an attacker can determine whether particular pages in the target site exist, by attempting to load them as resources and listening for the {{domxref("HTMLElement/error_event", "error")}} and {{domxref("HTMLElement/load_event", "load")}} events. If certain pages are only available to logged-in users, the attacker can determine whether the user is signed into the target site.

Author: -> https://github.com/mdn/content/pull/38977/commits/3566fa161c65e7ef7e107a794a9ddf03ed05ed17

All three attacks are deployed in the same way: the attacker crafts a page that implements the attack, then persuades the user to visit the page, for example by sending them an email or sharing a post containing the link. When the user visits the page the attack executes automatically.

In the rest of this section we'll describe these three attacks in a little more detail, to give you a concrete sense of how they work. Although the three attacks target quite different parts of the web platform, they have a common underlying cause: the extent to which the browser enables websites to connect to and interact with each other by mechanisms such as framing, loading subresources, or opening new windows.

> [!NOTE]
> For a more complete catalog of cross-site leaks, see the [XS-Leaks Wiki](https://xsleaks.dev/) and the [OWASP Cross-site Leaks Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/XS_Leaks_Cheat_Sheet.html).

> ### Leaking page existence using error events

Author: -> https://github.com/mdn/content/pull/38977/commits/3566fa161c65e7ef7e107a794a9ddf03ed05ed17

The attack relies on the ability of a website to load a resource from another site, for example by setting the `src` attribute of a {{htmlelement("script")}} element to the URL of the resource:

```js
const script = document.createElement("script");
script.src = "https://example.org/admin";
document.head.appendChild(script);
```

This results in an HTTP request to the `https://example.org/` website. If the request includes cookies that the site uses to identify users, and the page requested is only available to logged-in users, then the success or failure of the request reveals whether or not the user is logged in.

If the request fails, then the element fires an {{domxref("HTMLElement.error_event", "error")}} event. If the request succeeds, the element fires a {{domxref("HTMLElement.load_event", "load")}} event. By listening for these events, the attacker can discover whether the user is logged in.

```js
const url = "https://example.org/admin";
const script = document.createElement("script");

script.addEventListener("load", (e) => {
  console.log(`${url} exists`);
});

script.addEventListener("error", (e) => {
  console.log(`${url} does not exist`);

> });

Reviewer 5: I think you need to explain why script? Can you give an example of a response the web page may give that is an error/non-error depending on user state?

Author: > I think you need to explain why script? 

Isn't this explained at line 42:

> The attack relies on the ability of a website to load a resource from another site, for example by setting the `src` attribute of a {{htmlelement("script")}} element to the URL of the resource:

...or am I misunderstanding you? 

>  Can you give an example of a response the web page may give that is an error/non-error depending on user state?

I think I might need you to explain this a bit more for me. AIUI, in this scenario:
- if the user is logged in, then the page is returned, and `load`is fired
- if the user isn't logged in, then the server returns an error (e.g. 404) and `error` is fired

I sort of say that in 52, but is there something else/something more I need to say?

script.src = url;
document.head.appendChild(script);
```

An attacker may even be able to discover a user's ID, by iteratively trying to load pages to see if pages like `https://example.org/users/my_username` exist.

> 

Reviewer 5: I think it's important to note here (or above) that this is supposed to be a user-specific page, not "my_username" as a thing that is to be replaced. Maybe something like "example.org/personal-profile-settings" would be less nuanced?

Author: To check my understanding here: in this case we have a website that includes the the username in the URL for a page, so if the username is `fishfingers` and the user is logged in, then the page `https://example.org/users/fishfingers` exists. So an attacker can repeatedly try to load pages with test IDs, and if a page exits, then they know not only that the current user is logged in, but what their username is.

So (if this is right) maybe it's worth explaining it a bit more, with some version of that paragraph? Does that address your comment?

### Frame counting using window references

In a frame counting attack, the attacker finds out the number of frames currently loaded in the target page. In turn, that leaks information about the state of the target site, which may enable to attacker to learn, for example, whether the user is currently logged into the site.

If an attacker site gets a reference to a {{domxref("Window")}} object containing the target site, the attacker can count the number of frames in the target site by reading the {{domxref("Window.length", "window.length")}} property.

The attacker can get a `Window` object by calling {{domxref("Window.open()", "window.open()")}}:

```js
const target = window.open("https://example.org");
const frames = target.length;
```

Alternatively, the attacker can embed the target site in an {{htmlelement("iframe")}}, and retrieve the frame's {{domxref("HTMLIFrameElement.contentWindow", "contentWindow")}} property:

> 

Author: I'd prefer to keep defenses and attacks separate. I mean otherwise, why mention XFO here and not COOP? I do mention frame protection as a defense against this attack, in https://pr38977.review.mdn.allizom.net/en-US/docs/Web/Security/Attacks/XS-leaks#framing_protection.

In the attack described here, the attacker uses the [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/Guides/CSP) feature to detect whether a cross-site request was redirected.

- First, they create a page governed by a CSP that only allows {{htmlelement("iframe")}} elements to contain content from `https://example.org/admin`.

- Next, they add an event listener in the page that listens for the {{domxref("Document.securitypolicyviolation_event", "securitypolicyviolation")}} event.

- Finally, they create an {{htmlelement("iframe")}} element and set its `src` attribute to `https://example.org/admin`.

```html
<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="frame-src https://example.org/admin" />
  </head>
  <body>
    <script>
      document.addEventListener("securitypolicyviolation", () => {
        console.log("Page was redirected");
      });
      const frame = document.createElement("iframe");
      document.body.appendChild(frame);
      frame.src = "https://example.org/admin";
    </script>
  </body>
</html>

> ```

Author: Ah, could be. I don't think I tested this with a same-origin redirect, I tested with GitLab, which redirects gitlab.com to about.gitlab.com for not-logged-in visitors. So I suppose I could change this to admin.example.org?

Still I am surprised it wouldn't work same-origin. Doesn't `frame-src https://example.org/admin` forbid `https://example.org/login`, per https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy#host-source ?

> ```

Author: -> https://github.com/mdn/content/pull/38977/commits/944a50597215f4a80737aa2e47abd5e474310835

> ```

Reviewer 4: > Still I am surprised it wouldn't work same-origin. Doesn't frame-src https://example.org/admin forbid https://example.org/login

It does, but it onlly applies to the original URL not the one after redirects. This is to prevent information leaks if I remember correctly.

### Cross-Origin Opener Policy (COOP)

As we saw in the [frame-counting](#frame_counting_using_window_references) attack, another way to get a reference to the target's {{domxref("Window", "window")}} is as the return value of a call to {{domxref("Window.open()", "window.open()")}}:

```js
const target = window.open("https://example.com");
```

The {{httpheader("Cross-Origin-Opener-Policy")}} response header determines whether a document will be opened in the same {{glossary("Browsing context", "browsing context group")}} as the document that opened it.

If your server sends this header and sets it to any value except the default of `"unsafe-none"`, then if a document from a different origin tries to open your page using `window.open()`, your page will be loaded into a different browsing context group. Among other things, this means that the opener will not get a reference to the `window` object for your page, and will therefore not be able to use it in a frame counting attack.

### Defense summary checklist

> As we've seen, cross-site leaks include a range of attacks targeting different parts of the web platform: a single defense doesn't work against any of them. Indeed, some leaks, such as the one that exploits CSP to leak redirects, don't have any defenses yet.

Reviewer: I created a comment against a commit, but it doesn't seem to be visible here. Might be irrelevant

But while there is no way to stop CSP from revealing the redirect, you aren't forced to have redirects. You might, for example, use the same page with frames and different content for a logged in users.

In this guide we've outlined a few defenses that help to isolate your site from potential attackers, and we recommend implementing all of them:

- Use Fetch metadata to implement a resource isolation policy.
- Set the `SameSite` attribute for session cookies to `Strict` if you can, or `Lax` if you have to.
- Use the `frame-ancestors` CSP directive and the `X-Frame-Options` response header to prevent your site being embedded, or to control which sites can embed your site.
- Send the `Cross-Origin-Opener-Policy` response header to prevent other sites from accessing your `window` global object.

## See also

- [XS-Leaks Wiki](https://xsleaks.dev/) (xsleaks.dev)
- [Cross-site leaks Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/XS_Leaks_Cheat_Sheet.html) (OWASP)

> <section id="Quick_links">
> {{ListSubpages("/en-US/docs/Web/Security", "1", "0", "1")}}
> </section>

Reviewer 3: We can use the `security` sidebar now:

```suggestion
```

### Leaking redirects with a CSP

In some websites, the server will redirect a request, or not, based on whether the user is signed in (or has some special status on the site). For example, imagine a site which shows administrators a page at `https://example.org/admin`. If the user is not signed in, and requests this page, then the server might redirect them to `https://example.org/login`.

> 

Reviewer: ```suggestion
```
